core/libdeno/buffer.h

```diff
@@ -0,0 +1,153 @@
+// chromium //base. However Deno doensn't use that, so suppress that lint.
+#include <memory>
+#include <mutex>  // NOLINT
+#include <string>
+#include <unordered_map>
+#include <utility>
+
+#include "third_party/v8/include/v8.h"
+#include "third_party/v8/src/base/logging.h"
+
+namespace deno {
+
+template <typename T, typename S>
+static inline T BitMove(S* source) {
+  static_assert(sizeof(T) == sizeof(S),
+                "Source and target must have the same size.");
+  static const auto empty = T();
+  T moved;
+  memcpy(&moved, source, sizeof(T));
+  memcpy(source, &empty, sizeof(T));
+  return moved;
+}
+
+template <typename T>
+static inline T BitMove(T* source) {
+  return BitMove<T, T>(source);
+}
+
+class ArrayBufferAllocator : public v8::ArrayBuffer::Allocator {
+ public:
+  static ArrayBufferAllocator& global() {
+    static ArrayBufferAllocator global_instance;
+    return global_instance;
+  }
+
+  void* Allocate(size_t length) override { return new uint8_t[length](); }
+
+  void* AllocateUninitialized(size_t length) override {
+    return new uint8_t[length];
+  }
+
+  void Free(void* data, size_t length) override {
+    Unref(reinterpret_cast<uint8_t*>(data));
+  }
+
+ private:
+  friend class PinnedBufPin;
+
+  void Ref(void* data) {
+    std::lock_guard<std::mutex> lock(ref_count_map_mutex_);
+    // Note:
+    //  - `unordered_map::insert(make_pair(key, value))` returns the existing
+    //    item if the key, already exists in the map, otherwise it creates an
+    //    new entry with `value`.
+    //  - Buffers not in the map have an implicit reference count of one.
+    auto entry = ref_count_map_.insert(std::make_pair(data, 1)).first;
+    ++entry->second;
+  }
+
+  void Unref(void* data) {
+    {
+      std::lock_guard<std::mutex> lock(ref_count_map_mutex_);
+      auto entry = ref_count_map_.find(data);
+      if (entry == ref_count_map_.end()) {
+        // Buffers not in the map have an implicit ref count of one. After
+        // dereferencing there are no references left, so we delete the buffer.
+      } else if (--entry->second == 0) {
+        // The reference count went to zero, so erase the map entry and free the
+        // buffer.
+        ref_count_map_.erase(entry);
+      } else {
+        // After decreasing the reference count the buffer still has references
+        // left, so we leave the pin in place.
+        return;
+      }
+      delete[] reinterpret_cast<uint8_t*>(data);
+    }
+  }
+
+ private:
+  ArrayBufferAllocator() {}
+
+  ~ArrayBufferAllocator() {
+    // TODO(pisciaureus): Enable this check. It currently fails sometimes
+    // because the compiler worker isolate never actually exits, so when the
+    // process exits this isolate still holds on to some buffers.
+    // CHECK(ref_count_map_.empty());
+  }
+
+  std::unordered_map<void*, size_t> ref_count_map_;
+  std::mutex ref_count_map_mutex_;
+};
+
+class PinnedBufPin {
```

Comment:
This could use some explanatory comments. It’s not immediately clear to me what this does